File: user-service/src/main/kotlin/com/munchies/user/domain/port/TokenRepository.kt
package com.munchies.user.domain.port

import com.munchies.commons.Repository
import com.munchies.user.domain.model.Token
import com.munchies.user.domain.model.TokenId

interface TokenRepository : Repository<TokenId, Token> {
  fun isRevoked(token: String): Boolean
  fun setRevoked(token: String)
}

File: gateway-service/src/main/ts/infrastructure/adapter/outbound/token-repository.ts
export interface TokenRepository<TokenId> {
  add(id: TokenId): void;
  isRevoked(id: TokenId): boolean;
  revoke(id: TokenId): void;
}

class MemoryTokenRepository implements TokenRepository<string> {
  private readonly repository: Map<string, boolean> = new Map();

  add(id: string): void {
    this.repository.set(id, true);
  }
  isRevoked(id: string): boolean {
    return this.repository.get(id) ?? false;
  }
  revoke(id: string): void {
    this.repository.set(id, false);
  }
}

const repository = new MemoryTokenRepository();

export function getTokenRepository(): TokenRepository<string> {
  return repository;
}
